src/macros-by-example.adoc

```diff
@@ -462,25 +465,42 @@ types are verified:
 TODO: demonstrate splicing in TDL macro invocations
 
 
-=== Mapping Templates Over Streams
+=== Mapping Templates Over Streams: `*for*`
 
-Another way to produce a stream is via a mapping form. The `*for_each*` special form evaluates a
-template once for each value provided by a macro parameter.  Each time, the parameter name is
-shadowed and rebound to the next value on the stream.
+Another way to produce a stream is via a mapping form.  The `*for*` special form evaluates a
+template once for each value provided by a stream or streams.  Each time, a local variable is
+created and bound to the next value on the stream.
 
 [{mrk}]
 ----
 (*public* prices
-  [(*symbol* currency), (*number\...* amount)]
-  (*for_each* amount
-    (price amount currency)))
+  [(*symbol* currency), (*number\...* amounts)]
+  (*for* [(amt amounts)]
+    (price amt currency)))
 ----
+
+The list following `*for*` contains pairs of variable names and template expressions, so here
+each value from `amounts` is given the name `amt` before the `price` is expanded.
+
 ----
 (:prices GBP 10 9.99 12.)
   ⇒ {amount:10, currency:GBP} {amount:9.99, currency:GBP} {amount:12., currency:GBP}
 ----
 
-TODO: Consider the more-general binding form
+More than one stream can be iterated in parallel, and iteration terminates when any stream
+becomes empty.
```

Comment:
Not as specified.  The shorter streams could start producing void, but I think that would get quite complicated to handle properly and to debug.  Do you have a use case in mind?

Reply:
I have a sort-of use case—I can't think of a use case where we would want to _silently_ drop the end of the longer stream, and so being able to see that there are more elements remaining in stream could allow me (as a macro author) to raise an error when it occurs.

Is there any particular reason for it _not_ to be an error when the streams are different lengths?

Reply:
It's worth considering making uneven lengths an error. I'll cut an issue to track the idea.

FWIW I've been trying to avoid adding user-space error-handling mechanisms; I think its a sign that the language (or a macro author) is trying to do too much. It also opens the door to complexity necessary to do things like construct error messages, more-complex predicates, etc.